File: Tutorials/Codes/Modern_Control_Course/Chapter6/Lesson2/Chapter6_Lesson2.py
import numpy as np
import scipy.linalg as la

# Optional (if available): pip install control
try:
    import control as ct
except Exception:
    ct = None

A = np.array([[0.0, 1.0],
              [-2.0, -3.0]])
B = np.array([[0.0],
              [1.0]])
C = np.array([[1.0, 1.0]])
D = np.array([[0.0]])

# Poles from A (internal candidate poles)
poles_A = la.eigvals(A)
print("eig(A) =", poles_A)

# SISO transmission zeros via generalized eigenvalues of a Rosenbrock pencil
# P(s) = s*E - F, with
# E = [[I, 0],
#      [0, 0]]
# F = [[A, B],
#      [-C, -D]]
n = A.shape[0]
E = np.block([[np.eye(n), np.zeros((n,1))],
              [np.zeros((1,n)), np.zeros((1,1))]])
F = np.block([[A, B],
              [-C, -D]])

# Generalized eigenvalues solve F v = lambda E v (lambda are finite zeros + possibly inf)
alpha, beta = la.eig(F, E, right=False, left=False, overwrite_a=False, overwrite_b=False, homogeneous_eigvals=True)
# Convert homogeneous eigenvalues alpha/beta to lambda, filtering beta=0 (infinite eigenvalues)
zeros = []
for a, b in zip(alpha, beta):
    if abs(b) > 1e-12:
        zeros.append(a / b)
zeros = np.array(zeros, dtype=complex)
print("zeros (finite generalized eig) =", zeros)

# Validate by forming transfer function (if python-control is installed)
if ct is not None:
    sys = ct.ss(A, B, C, D)
    tf_sys = ct.tf(sys)
    print("Transfer function (unsimplified) =", tf_sys)
    print("Poles from tf =", ct.poles(tf_sys))
    print("Zeros from tf =", ct.zeros(tf_sys))

File: Tutorials/Codes/Modern_Control_Course/Chapter6/Lesson2/Chapter6_Lesson2.cpp
#include <iostream>
#include <vector>
#include <complex>

#include <Eigen/Dense>
#include <Eigen/Eigenvalues>

int main() {
  using Eigen::MatrixXd;
  using Eigen::VectorXcd;

  MatrixXd A(2,2); A << 0, 1,
                         -2, -3;
  MatrixXd B(2,1); B << 0,
                         1;
  MatrixXd C(1,2); C << 1, 1;
  MatrixXd D(1,1); D << 0;

  // Poles: eigenvalues of A
  Eigen::EigenSolver<MatrixXd> esA(A);
  VectorXcd poles = esA.eigenvalues();
  std::cout << "eig(A) poles:\n" << poles << "\n\n";

  // Build Rosenbrock pencil for SISO:
  // P(s) = [[sI-A, -B],[C, D]] = s*E - F
  // E = [[I, 0],[0, 0]], F = [[A, B],[-C, -D]]
  int n = 2;
  MatrixXd E = MatrixXd::Zero(n+1, n+1);
  E.block(0,0,n,n) = MatrixXd::Identity(n,n);

  MatrixXd F = MatrixXd::Zero(n+1, n+1);
  F.block(0,0,n,n) = A;
  F.block(0,n,n,1) = B;
  F.block(n,0,1,n) = -C;
  F(n,n) = -D(0,0);

  // Generalized eigenvalues: F v = lambda E v
  Eigen::GeneralizedEigenSolver<MatrixXd> ges(F, E);
  VectorXcd lambdas = ges.eigenvalues();

  std::cout << "Generalized eigenvalues (finite zeros appear among these):\n";
  for (int i=0; i<lambdas.size(); ++i) {
    std::complex<double> lam = lambdas(i);
    // Filter huge magnitudes as "infinite" placeholders (heuristic)
    if (std::abs(lam) < 1e8) {
      std::cout << "  " << lam << "\n";
    }
  }
  return 0;
}
